libcrux-ml-dsa/src/simd/portable/encoding/commitment.rs

```diff
@@ -1,40 +1,158 @@
+ forall (i: nat {i < 8 * 4}). inp i == out i)
+")
+})]
+pub fn serialize_4(simd_unit: &Coefficients, serialized: &mut [u8]) {
+    // The commitment has coefficients in [0,15] => each coefficient occupies
+    // 4 bits.
+    cloop! {
+        for (i, coefficients) in simd_unit.values.chunks_exact(2).enumerate() {
+            hax_lib::loop_invariant!(|i: usize| {fstar!(r"
+            Seq.length $serialized == 4 /\ (
+              let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+              let out = bit_vec_of_int_t_array #U8 #(mk_usize 4) $serialized 8 in
+              forall (n: nat {n < v i * 8}). out n == inp n
+            )")});
+            let r0 = encode_4(coefficients);
+            serialized[i] = r0;
+            hax_lib::fstar!(r"
+                let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+                let out = bit_vec_of_int_t_array #U8  #(mk_usize 4) $serialized 8 in
+                introduce forall (n:nat{n < 8}). inp (v i * 8 + n) == out (v i * 8 + n)
+                with
+                  (calc (==) {
+                          inp (v i * 8 + n);
+                    == {} get_bit (Seq.index ${simd_unit.values} ((v i * 8 + n) / 4)) (sz ((v i * 8 + n) % 4));
+                    == { Math.Lemmas.division_addition_lemma n 4 (v i * 2) }
+                          get_bit (Seq.index ${simd_unit.values} (v i * 2 + n / 4)) (sz (n % 4));
+                    == {} get_bit (Seq.index $coefficients (n / 4)) (sz (n % 4));
+                    == {} bit_vec_of_int_t_array #I32 #(mk_usize 2) $coefficients 4 n;
+                    == {} out (v i * 8 + n);
+                  })
+            ");
         }
+    }
+    ()
+}
+
+#[inline(always)]
+#[hax_lib::requires(fstar!(r"Seq.length ${serialized} == 6 /\ (forall i. bounded (Seq.index ${simd_unit.values} i) 6)"))]
+#[hax_lib::ensures(|out| {
+    let serialized_future = future(serialized);
+    fstar!(r"
+Seq.length ${serialized_future} == Seq.length ${serialized} /\
+(let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 6 in
+ let out = bit_vec_of_int_t_array #U8  #(mk_usize 6) ${serialized_future} 8 in
+ forall (i: nat {i < 8 * 6}). inp i == out i)
+")
+})]
+pub fn serialize_6(simd_unit: &Coefficients, serialized: &mut [u8]) {
+    // The commitment has coefficients in [0,43] => each coefficient occupies
+    // 6 bits.
+    for (i, coefficients) in simd_unit.values.chunks_exact(4).enumerate() {
+        hax_lib::loop_invariant!(|i: usize| {
+            fstar!(
+                r"
+            Seq.length $serialized == 6 /\ (
+              let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 6 in
+              let out = bit_vec_of_int_t_array #U8 #(mk_usize 6) $serialized 8 in
+              (forall (n: nat {n < v i * 24}). out n == inp n))"
+            )
+        });
+        let (r0, r1, r2) = encode_6(coefficients);
```

Comment:
Why doesn't `encode_6` (and `encode_4`) take a mutable reference to `serialized` to write right into it?

Reply:
Stateful updates make both proofs and proof statements more complicated: we need to state that nothing changes but the updated value(s), which implies a universally quantifier
So here I prefer to cut the proof in two: the serialization problem (ie prove correctness for shifts & friends) and then the loopy-imperative-update things.

Reply:
Ok, if there's a good reason to do this, it's fine. But please add an explanation into comments why it is done this way, as it's not what we'd usually write in Rust.

Reply:
I think we should be able to do a version where the `encode` functions mutate the output array instead of returning a tuple. I am putting this improvement for the followup issue: https://github.com/cryspen/libcrux/issues/777